crates/uv-trampoline/src/bounce.rs

```diff
@@ -114,34 +126,156 @@ fn executable_filename() -> CString {
+        );
+
+        let mut read_bytes = 0u32;
+
+        expect_result(
+            unsafe {
+                ReadFile(
+                    file_handle,
+                    buffer.as_mut_ptr() as *mut _,
+                    bytes_to_read,
+                    &mut read_bytes,
+                    null_mut(),
+                )
+            },
+            0,
+            || String::from("Failed to read the executable file"),
+        );
+
+        // Truncate the buffer to the actual number of bytes read.
+        buffer.truncate(read_bytes as usize);
+
+        if !buffer.ends_with(&MAGIC_NUMBER) {
+            eprintln!("Magic number 'UVUV' not found at the end of the file. Did you append the magic number, the length and the path to the python executable at the end of the file?");
+            exit_with_status(1);
+        }
+
+        // Remove the magic number
+        buffer.truncate(buffer.len() - MAGIC_NUMBER.len());
+
+        let path_len = match buffer.get(buffer.len() - PATH_LEN_SIZE..) {
+            Some(path_len) => {
+                let path_len = u32::from_le_bytes(path_len.try_into().unwrap_or_else(|_| {
+                    eprintln!("Slice length is not equal to 4 bytes");
+                    exit_with_status(1)
+                }));
+
+                if path_len > MAX_PATH_LEN {
+                    eprintln!("Only paths with a length up to 32KBs are supported but the python path has a length of {}.", path_len);
+                    exit_with_status(1);
+                }
+
+                // SAFETY: path len is guaranteed to be less than 32KBs
+                path_len as usize
+            }
+            None => {
+                eprintln!("Python executable length missing. Did you write the length of the path to the Python executable before the Magic number?");
+                exit_with_status(1);
+            }
+        };
+
+        // Remove the path length
+        buffer.truncate(buffer.len() - PATH_LEN_SIZE);
+
+        if let Some(path_offset) = buffer.len().checked_sub(path_len) {
+            buffer.drain(..path_offset);
+            buffer.push(b'\0');
+
+            break CString::from_vec_with_nul(buffer).unwrap_or_else(|_| {
+                eprintln!("Python executable path is not correctly null terminated.");
+                exit_with_status(1)
+            });
+        } else {
+            // SAFETY: Casting to u32 is safe because `path_len` is guaranteed to be less than 32KBs,
+            // MAGIC_NUMBER is 4 bytes and PATH_LEN_SIZE is 4 bytes.
+            bytes_to_read = (path_len + MAGIC_NUMBER.len() + PATH_LEN_SIZE) as u32;
+
+            if i64::from(bytes_to_read) > file_size {
+                eprintln!("The length of the python executable path exceeds the file size. Verify that the path length is appended to the end of the launcher script as a u32 in little endian.");
+                exit_with_status(1);
+            }
```

Comment:
At first I was unsure if this was limiting heap memory since `bytes_to_read` was still being calculated based on `path_len`, but I see above that `path_len` is limited to a maximum. So I think that's right.